### ISSUE TITLE
Not showing elements in one of our model

### ISSUE THREAD

**Issue by Chris-12** [opensourceBIM/BIMserver#1217]
Hi all,
we have a model for which important parts are not rendered in BIMvie.ws, nor on our javascript web page.
If we load it with e.g. bimVision, these parts are correctly displayed.
In BIMVie.ws, we have double checked that all categories (types, layers, classifications) are all set on (shown),
and that there is not error while checkin the model in, nor while loading it in BIMvie.ws.
This model is an ifc (ifc2x3tc1) file, of about 800Mo.
We use BimServer release 182, on a windows server2012 with 32Go ram.
We cannot publicly share this file, but in private yes.

Can anyone help, or guide us ?
Regards,
Chris

**Comment by hlg** [comment]
When you check for upload errors, do you also look at the geometry generation report (under extended data)?

**Comment by hlg** [comment]
I was able to reproduce the issue with your model. The geometry report does not show errors, but I could identify one or two jobs (depending on the configuration) that report a number of 0 triangles and a negative duration. Do you also see this?

The issue does not appear when offending elements are uploaded and geometry is generated for them in isolation. Thus we don't have an issue within IfcOpenShell. I assume that the problem is the size of geometries, since those missing elements have very large geometries and moreover, a bunch of those large geometries is bundled in one job. This seems to lead to a situation where IfcOpenShell terminates in a way that is not handled by BIMserver or even cannot be handled. Maybe it is a memory issue, where RAM is not enough to cover JVM plus IfcOpenShell requirements. You could try to reduce JVM heap (such that it is still enough in order for BIMserver to handle the large file) and see if this then leaves enough for IfcOpenShell to do the job. Not sure if this would work though.

I have to further investigate if this is indeed the root cause. If so, I would a) try to introduce a proper error message, but this might be impossible and b) try to adjust the boxing of products into jobs to also consider product geometry, probably by introducing an additional limit on number of geometry entities  besides the number of products per job.

**Comment by zaqifathis** [comment]
Is this issue still happened with BIMserver 1.5.185?

**Comment by Chris-12** [comment]
Hi I am going to test this with release 186...

**Comment by Chris-12** [comment]
Well, the missing elements are now correctly showing. This is a good thing.
But it seems that we cannot select elements inside the former missing element. Actually, when I click on an element, an other part of it is selected.

**Comment by hlg** [comment]
Great. What do you mean with "inside"?

**Comment by Chris-12** [comment]
There was an electrical set (electric circuit breaker probably) containing many elements and this whole et was not showing. OK.
Now that it is showing with its contained elements (electical parts), I cannot click and select the different parts because selection seems to be random (actually with the bimviews).

**Comment by hlg** [comment]
I see what you mean. This seems to be a viewer issue. Both the storage in BIMserver as well as geometry generation are fine, there is no issue. In IFC, the respective element is actually just one entity (no parts) and you would not expect to select distinct parts of it, but rather the whole thing when you click on it.

BIMvie.ws contains an older version of BIMsurfer as a viewer which in turn uses scene.js as library. The entity in question has a very large geometry with many triangles. The geometry serializer responsible for feeding the BIMsurfer viewer, splits geometry into parts if there are more than 16.389 index values (5.463 triangles). I don't know where that number comes from - most likely it is related to some limitation of the scene graph library in the viewer, scene.js. (I tried to disable the splitting in the serializer and indeed it will then only show a fragment of the geometry.) Now those several groups of triangles constitute multiple nodes in the scene graph and you can select only one at a time. Thus when you click somewhere, one group is highlighted and the highlighting cycles through the groups of triangles in a more or less stable order.

This particular version of the BIMsurfer is not maintained anymore and technically we don't provide updates and fixes for this viewer, also because the scene.js library is unmaintained as well. If someone wants to go ahead and modify the viewer code for consistent hightlighting of the whole splitted entity at once when clicking, you could start here: https://github.com/opensourceBIM/BIMsurfer-before2019/blob/V1/api/Control/ClickSelect.js.

As we cannot do anything about this right now, I would close this issue.

**Comment by Chris-12** [comment]
I understand. I have implemented BimSurfer 3 a few years ago, because older versions of BimSurfer/scene.js were kind of poor displaying models.
It is a bad news that no viewer is currently maintained up-to-date with currently releasing BimServer versions.
If scene.js is about to die, what open source library viewer would fit at most the structure of BIM Server, according to you ?

**Comment by hlg** [comment]
Do you see the same highlighting issue with BIMsurfer 3 as well? With that version of the viewer, you can configure the splitting via `splitGeometry` loader setting. By default splitting is turned of, so theoretically this problem should not occur:

https://github.com/opensourceBIM/BIMsurfer/blob/414695b76edd25b9a5cc7334e71fd7b279067614/viewer/defaultsettings.js#L76

**Comment by Chris-12** [comment]
Well no, actually with BIMSurfer3 when I click on the electrical set, the whole thing is selected. It seems that there is no mean to select elements inside the set, whether I set splitGeometry on true or false.

**Comment by hlg** [comment]
Yes, that is how it is supposed to be, because in the electrical object is essentially modelled as "one thing". The splitting is only done on geometry level (into random sets of triangles) because the geometry is too big for the old BIMsurfer viewer.

**Comment by hlg** [comment]
Seems to be resolved.
